File: src/text.zig
//! An 8x16 bitmap font and a blitter, shared by the shm apps (which write u32
//! pixels) and the GL apps (which rasterise into an R8 coverage texture).
//!
//! assets/font8x16.bin is the ASCII range of Terminus (OFL-1.1), extracted
//! from Lat2-Terminus16.psfu as 95 glyphs of 16 bytes, one bit per pixel,
//! MSB leftmost.
const font = @embedFile("font");

pub const GLYPH_W = 8;
pub const GLYPH_H = 16;

pub fn glyph(ch: u8) ?*const [GLYPH_H]u8 {
    if (ch < 0x20 or ch >= 0x7f) return null;
    return font[(@as(usize, ch) - 0x20) * GLYPH_H ..][0..GLYPH_H];
}

/// Blit `s` at pixel (x, y) into a `w` x `h` buffer of `P`, `scale`x magnified.
/// Anything off the right or bottom edge is clipped.
pub fn draw(
    comptime P: type,
    dst: []P,
    w: usize,
    h: usize,
    x: usize,
    y: usize,
    scale: usize,
    colour: P,
    s: []const u8,
) void {
    var cx = x;
    for (s) |ch| {
        defer cx += GLYPH_W * scale;
        if (cx + GLYPH_W * scale > w) return;
        const g = glyph(ch) orelse continue;
        for (g, 0..) |bits, gy| {
            if (bits == 0) continue;
            for (0..GLYPH_W) |gx| {
                if (bits >> @intCast(7 - gx) & 1 == 0) continue;
                for (0..scale) |sy| {
                    const py = y + gy * scale + sy;
                    if (py >= h) continue;
                    for (0..scale) |sx| dst[py * w + cx + gx * scale + sx] = colour;
                }
            }
        }
    }
}

/// Width in pixels of `s` at `scale`.
pub fn widthOf(s: []const u8, scale: usize) usize {
    return s.len * GLYPH_W * scale;
}

File: src/ui/text.c
#include "text.h"

#include "font8x16.h"

static const uint8_t *glyph(unsigned char ch)
{
    if (ch < 0x20 || ch >= 0x7f)
        return NULL;
    return font8x16 + ((size_t)ch - 0x20) * TEXT_GLYPH_H;
}

void text_draw(uint8_t *dst, size_t w, size_t h, size_t x, size_t y, size_t scale,
               uint8_t value, const char *text)
{
    size_t cx = x;
    for (const unsigned char *p = (const unsigned char *)text; *p != '\0';
         p++, cx += TEXT_GLYPH_W * scale) {
        if (cx + TEXT_GLYPH_W * scale > w)
            return;
        const uint8_t *bitmap = glyph(*p);
        if (bitmap == NULL)
            continue;
        for (size_t gy = 0; gy < TEXT_GLYPH_H; gy++) {
            const uint8_t bits = bitmap[gy];
            if (bits == 0)
                continue;
            for (size_t gx = 0; gx < TEXT_GLYPH_W; gx++) {
                if (((bits >> (7 - gx)) & 1) == 0)
                    continue;
                for (size_t sy = 0; sy < scale; sy++) {
                    const size_t py = y + gy * scale + sy;
                    if (py >= h)
                        continue;
                    for (size_t sx = 0; sx < scale; sx++)
                        dst[py * w + cx + gx * scale + sx] = value;
                }
            }
        }
    }
}
